keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+
+### Ingress Controller Operation
+
+The [Kubernetes ingress feature](https://kubernetes.io/docs/concepts/services-networking/ingress/) relies on the use of an ingress controller. The two "reference" ingress controllers that are considered here are the [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) and the [NGINX ingress controller](https://github.com/kubernetes/ingress-nginx/blob/master/README.md#nginx-ingress-controller).
+
+#### GCE Ingress Controller: Does Not Support External, Dual-Stack Operation
+It appears that [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) support is limited to IPv4-only access (QUESTION: Is this true???) from outside of the cluster. Therefore, GCE ingress controllers will not be capable of supporting external, dual-stack access to services running in a dual-stack Kubernetes cluster.
+
+#### NGINX Ingress Controller - Dual-Stack Support for Bare Metal Clusters
+The [NGINX ingress controller](https://github.com/kubernetes/ingress-nginx/blob/master/README.md#nginx-ingress-controller) should provide dual-stack external access to Kubernetes services that are hosted on baremetal clusters, with little or no changes.
+
+- Dual-stack external access to NGINX ingress controllers is not supported with GCE/GKE or AWS cloud platforms.
+- NGINX ingress controller needs to be run on a pod with dual-stack external access (needs to be on host network, or can be run directly on a pod network???)
+- On the load balancer (internal) side of the NGINX ingress controller, dual-stack access to Kubernetes service backends requires the resolver to be configured with 'ipv6=on' (which is default, see the ["Configuring HTTP Load Balancing Using DNS" section of the NGINX ingress controller admin guide](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-load-balancer/#configuring-http-load-balancing-using-dns).
+- On the load balancer (internal) side of the NGINX ingress controller, the controller will load balance to backend service pods on a per-address basis, rather than load balancing on a per-dual-stack-endpoint basis. For example, if a given backend pod has both an IPv4 and an IPv6 address, the ingress controller will treat the IPv4 and IPv6 address endpoints as separate load-balance targets. (Changing this behavior would require upstream changes to the NGINX ingress controller.)
+- Ingress access can cross IP families. For example, an incoming L7 request that is received via IPv4 can be load balanced to an IPv6 endpoint address in the cluster, and vice versa. 
+
+### Load Balancer Operation ???
+
+### Network Policy Considerations ???
+
+### Cloud Provider Plugins Considerations ???
+
+### kubeadm Support
+\<TBD\>
+
+### End-to-End Test Support
+\<TBD\>
+
+### Risks and Mitigations
+\<TBD\>
+
+## Graduation Criteria
+\<TBD\>
+
+## Implementation History
+\<TBD\>
+
+## Drawbacks [optional]
+\<TBD\>
+
+## Alternatives [optional]
```

Comment:
I think this would be good to have up top, where we discuss the motivation, and then the ways to address the needs and the selection chosen.